Disable delay_optimizer_creation in `Trainer` to support fsdp2

byi8220 opened the pull request:
# What does this PR do?

In order to get Trainer support for FSDP2 in `accelerate`, we have to pass the model and optimizer into `Accelerator.prepare()` at the same time (https://github.com/huggingface/accelerate/pull/3394/files#r2017637611).

**Note:** This may not be sufficient for full FSDP2 support in Trainer. This PR might be scrapped and replaced with something more complete if some issues arise.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


### Testing

2 unit tests were added to `tests/fsdp/test_fsdp.py`

1. `test_fsdp2_cpu_offloading`, which matches `test_fsdp_cpu_offloading`, but is **skipped** since the second test appears to be non functioning
2. `test_accelerate_fsdp2_integration` which matches `test_training_and_can_resume_normally` with SHARDED_STATE_DICT (as this is the only config that can run accelerate), and passes.

This will not be caught by CI, as it depends on an unreleased feature, but was instead manually run with the command `RUN_SLOW=1 pytest tests/fsdp/test_fsdp.py `
```
# RUN_SLOW=1 pytest tests/fsdp/test_fsdp.py 

tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_full_shard_bf16 PASSED                                                                                                                                             [  4%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_full_shard_fp16 PASSED                                                                                                                                             [  8%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_shard_grad_op_bf16 PASSED                                                                                                                                          [ 13%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_shard_grad_op_fp16 PASSED                                                                                                                                          [ 17%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_with_cpu_offload_0_bf16 PASSED                                                                                                                                     [ 21%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_with_cpu_offload_1_fp16 PASSED                                                                                                                                     [ 26%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_with_gradient_accumulation_full_shard_bf16 PASSED                                                                                                                  [ 30%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_with_gradient_accumulation_full_shard_fp16 PASSED                                                                                                                  [ 34%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_with_gradient_accumulation_shard_grad_op_bf16 PASSED                                                                                                               [ 39%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_basic_run_with_gradient_accumulation_shard_grad_op_fp16 PASSED                                                                                                               [ 43%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_cpu_offloading SKIPPED (FSDP CPU offloading script not found!)                                                                                                          [ 47%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_training_and_can_resume_normally_FULL_STATE_DICT PASSED                                                                                                                      [ 52%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_training_and_can_resume_normally_SHARDED_STATE_DICT PASSED                                                                                                                   [ 56%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_accelerate_fsdp2_integration PASSED                                                                                                                                          [ 60%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp2_cpu_offloading SKIPPED (FSDP 2 CPU offloading script not found!)                                                                                                       [ 65%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_full_shard_bf16 PASSED                                                                                                                                           [ 69%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_full_shard_fp16 PASSED                                                                                                                                           [ 73%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_shard_grad_op_bf16 PASSED                                                                                                                                        [ 78%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_shard_grad_op_fp16 PASSED                                                                                                                                        [ 82%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_transformers_auto_wrap_full_shard_bf16 PASSED                                                                                                                    [ 86%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_transformers_auto_wrap_full_shard_fp16 PASSED                                                                                                                    [ 91%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_transformers_auto_wrap_shard_grad_op_bf16 PASSED                                                                                                                 [ 95%]
tests/fsdp/test_fsdp.py::TrainerIntegrationFSDP::test_fsdp_config_transformers_auto_wrap_shard_grad_op_fp16 PASSED                                                                                                                 [100%]
```

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @zach-huggingface and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc @zach-huggingface
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

byi8220 committed:
github why you do this

byi8220 committed:
fix

byi8220 committed:
Merge branch 'huggingface:main' into trainer-fsdp2-2

byi8220 committed:
Merge branch 'huggingface:main' into trainer-fsdp2-2

S1ro1 reviewed (approved):
LGTM. cc @SunMarc for double-check on the added tests.

byi8220 committed:
make fixup

byi8220 committed:
disable cpu offload test

byi8220 committed:
fixup

SunMarc commented on tests/fsdp/test_fsdp.py:
```diff
@@ -134,6 +134,14 @@ def _parameterized_custom_name_func(func, param_num, param):
     return f"{func.__name__}_{param_based_name}"
 
 
+if is_accelerate_available():
+    from accelerate.accelerator import Accelerator
+
+    is_fsdp2_available = hasattr(Accelerator, "is_fsdp2")
+else:
+    is_fsdp2_available = False
+
+
```
try to do something similar to what we have with fsdp with `require_fsdp_version` and we can also check accelerate version too.

byi8220 replied:
>try to do something similar to what we have with fsdp with require_fsdp_version 

Can do, I think fsdp2 needs torch 2.6.0 or higher

>and we can also check accelerate version too.

Oh lol, this was literally just released as https://github.com/huggingface/accelerate/releases/tag/v1.6.0 an hour ago. I can change this to need accelerate>=1.6.0

byi8220 replied:
Done. Looks a bit strange imo, but such is version/dep juggling.

SunMarc reviewed (commented):
Thanks ! Left a couple of comments. It would be nice to check that all the fsdp tests works with fsdpv2 also @S1ro1 but that can be done in a follow-up PR

byi8220 commented on src/transformers/trainer.py:
```diff
@@ -2313,6 +2313,11 @@ def _inner_training_loop(
 
         delay_optimizer_creation = is_sagemaker_mp_enabled() or self.is_fsdp_xla_enabled or self.is_fsdp_enabled
 
+        # Can't delay optimizer creation when using FSDP2: https://github.com/huggingface/accelerate/blob/3f636d626063ffcf9a337c7d3624d61b7d187d59/src/accelerate/accelerator.py#L1404
+        is_fsdp2 = (self.accelerator is not None) and getattr(self.accelerator, "is_fsdp2", False)
+        if is_fsdp2:
+            delay_optimizer_creation = False
+
```
I can change that but may I ask why that would be better?

IIUC, [Accelerator.is_fsdp2](https://github.com/huggingface/accelerate/blob/83e09a93314b20efbe10f1545a1b1679c0b06862/src/accelerate/accelerator.py#L714-L716) delegates to [AcceleratorState.is_fsdp2](
https://github.com/huggingface/accelerate/blob/83e09a93314b20efbe10f1545a1b1679c0b06862/src/accelerate/state.py#L1059-L1061) which ultimately is checking its plugin.fsdp_version. Feels like it makes sense for Accelerator's users to check from the outside rather than prying in, but it's not a big deal and I can change this if it makes more sense

SunMarc replied:
We already do that for `is_fsdp_enabled`, so I prefer to do the same here as well.

byi8220 replied:
Rewrote it to be similar. We trust `self.is_fsdp_enabled` implies the existence of `self.accelerator.state.fsdp_plugin`

byi8220 committed:
Merge branch 'huggingface:main' into trainer-fsdp2-2

byi8220 committed:
tmp reworks

byi8220 committed:
git branch movement

byi8220 committed:
make fixup

byi8220 committed:
add require_fsdp_v2_version

byi8220 committed:
dep issues

byi8220 commented:
@SunMarc I've addressed your comments and the test suite passes on `torch==2.6.0`

I'm a bit worried about how this interacts with `torch` version 2.5.1 (seems FSDP2 wasn't "officially" supported until 2.6.0), but testing on `torch==2.5.1` on my machine, the `main` branch doesn't even pass

byi8220 committed:
Merge branch 'main' into trainer-fsdp2-2

SunMarc reviewed (approved):
Thanks ! make sure to update your ruff version to fix the quaility CI test

SunMarc commented:
cc @S1ro1 for FSDP2_PYTORCH_VERSION comment. Maybe we need to update it to 2.6.0 ?

byi8220 committed:
Merge branch 'main' into trainer-fsdp2-2

byi8220 committed:
Merge branch 'huggingface:main' into trainer-fsdp2-2

byi8220 committed:
update ruff and fixup

byi8220 committed:
Merge branch 'main' into trainer-fsdp2-2

byi8220 committed:
Merge branch 'main' into trainer-fsdp2-2

byi8220 committed:
Merge branch 'main' into trainer-fsdp2-2
